### PR TITLE
Changes traffic_wccp to use YAML as configuration format

### PR DESCRIPTION
**Pull request by randall**

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
Why was the line number for errors taken out?

**Comment by randall** [comment]
[approve ci freebsd]

**Comment by randall** [comment]
> Why was the line number for errors taken out?

They're  now back!

File: src/wccp/WccpConfig.cc [hunk 1]
```diff
@@ -56,7 +55,9 @@ using namespace wccp;
 std::vector<uint32_t> Seed_Router;
 
 // Names used for various elements and properties.
-static const char *const SVC_NAME = "service";
+static const char *const WCCP_NAME = "wccp";
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: These should be `std::string`, otherwise one is constructed on every lookup.

File: src/wccp/WccpConfig.cc [hunk 2]
```diff
@@ -179,47 +137,35 @@ Service_Type_Defaulted(wccp::ServiceGroup::Type type, int line)
 }
 
 ts::Errata::Message
-Service_Type_Invalid(ts::ConstBuffer const &text, int line)
+Service_Type_Invalid(const std::string &text, int line)
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: Use `string_view` in these case. Otherwise, as with YAMLCPP, use of a `string_view` or literal string will require the construction of a `std::string`. The C++ stream operators handle `string_view` correctly.

Comment (SolidWallOfCode) [reply src/wccp/WccpConfig.cc]: I'll note that `ts::string_view` and `ts::TextView` are direct descendants of `ts::ConstBuffer`.

File: src/wccp/WccpConfig.cc [hunk 3]
```diff
@@ -257,14 +186,6 @@ Svc_Ports_Too_Many(int line, int n)
   return ts::Errata::Message(14, LVL_INFO, out.str());
 }
 
-ts::Errata::Message
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: Why is this removed? What if someone puts "Boo, Randal" as a service port value?

File: src/wccp/WccpConfig.cc [hunk 4]
```diff
@@ -367,342 +288,339 @@ operator==(const char *text, ts::ConstBuffer const &b)
...
-      if (ts::config::IntegerValue == prop.getType()) {
-        x = atoi(prop.getText()._ptr);
-        if (0 <= x && x <= 255)
-          svc_info.setSvcId(x);
-        else
-          zret.push(Svc_Prop_Out_Of_Range(SVC_PROP_ID, prop, x, 0, 255));
+    YAML::Node propId = svc_cfg[SVC_PROP_ID];
+    if (propId) {
+      int x = propId.as<int>();
+      if (0 <= x && x <= 255) {
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: What value does `as<int>("ganchovy")` yield? 0? In which case any random string will appear to be valid.

File: src/wccp/WccpConfig.cc [hunk 5]
```diff
@@ -367,342 +288,339 @@ operator==(const char *text, ts::ConstBuffer const &b)
...
-          else
-            zret.push(Svc_Ports_None_Valid(src_line));
+          } else {
+            zret.push(Svc_Ports_None_Valid(ports.Mark().line));
+          }
         } else {
-          zret.push(Prop_Invalid_Type(prop, ts::config::ListValue));
+          // port is a scalar
+          int x = ports.as<int>();
+          svc_info.setPort(0, x); // ??? start at 0? same with sidx
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: Heh. You could move the `sidx` declaration to just under the `else` and use `svc_info.setPort(sidx++, x);` here as well.

File: src/wccp/WccpConfig.cc [hunk 6]
```diff
@@ -712,86 +630,96 @@ CacheImpl::loadServicesFromFile(const char *path)
...
-    if ((prop = svc_cfg[SVC_PROP_PROC]).hasValue()) {
-      if (ts::config::StringValue == prop.getType()) {
-        svc.setProcName(prop.getText());
-      } else {
-        zret.push(Prop_Invalid_Type(prop, ts::config::StringValue));
-      }
+    YAML::Node proc_name = svc_cfg[SVC_PROP_PROC];
+    if (proc_name) {
+      std::string text = proc_name.as<std::string>();
+      svc.setProcName(ts::ConstBuffer(text.c_str(), text.size()));
```
Comment (SolidWallOfCode) [inline src/wccp/WccpConfig.cc]: As long as you're cleaning up, get rid of `ConstBuffer` entirely. Use `string_view`. If you change `svc.setProcName` to take a `string_view` then you can simply pass `text` from here.

File: src/wccp/WccpConfig.cc (hunk 4 above)
Comment (randall) [reply src/wccp/WccpConfig.cc]: An exception is thrown. error at line X, column X: bad conversion

File: src/wccp/WccpConfig.cc (hunk 3 above)
Comment (randall) [reply src/wccp/WccpConfig.cc]: The parser will catch it and will generate it's own standard messaging. And in the future, something better(canned yams?) will  catch it before it even happens (at validation).

**Comment by randall** [comment]
Issues addressed (I think)!

Go easy on me, I'm a string_view n00b.

**Comment by SolidWallOfCode** [comment]
I'll send you a PGE button :-).

**Comment by randall** [comment]
@SolidWallOfCode Is changes ok now?

**Comment by zwoop** [comment]
@oknet Can you take a look at this too please, since you are the primary user of WCCP.

**Comment by randall** [comment]
@oknet Do you have any concerns here?
